File: src/main/js/formulas/info/TYPE.ts
import { isNull } from "../../utils/Other";
import { F7Exception } from "../../errors/F7Exception";
import { Grid } from "../../models/common/Grid";
import { SheetColumnRowKey } from "../../models/common/SheetColumnRowKey";
import { Complex } from "../../models/common/Types";
import { Converters } from "../../utils/Converters";
import { AbstractFormula } from "../AbstractFormula";
import { FormulaName } from "../FormulaName";

export class TYPE extends AbstractFormula {
  static SELF: TYPE = new TYPE();
  NAME = FormulaName.TYPE;

  internal(origin: SheetColumnRowKey, ...values: Array<Complex>) {
    AbstractFormula.checkLength(values.length, 1, this.NAME);
    const value = this.collateralLookup(origin, values[0]);
    if (isNull(value) || typeof value === "number") {
      return 1;
    }
    if (typeof value === "string") {
      return 2;
    }
    if (typeof value === "boolean") {
      return 4;
    }
    if (value instanceof F7Exception) {
      return 16;
    }
    if (value instanceof Grid) {
      const grid = Converters.castAsGrid(value);
      if (grid.isSingle()) {
        return TYPE.SELF.run(origin, grid.get(0, 0));
      }
      return 64;
    }
    return 128;
  }
}

File: src/main/java/io/protobase/f7/formulas/info/TYPE.java
package io.protobase.f7.formulas.info;

import io.protobase.f7.errors.F7Exception;
import io.protobase.f7.formulas.AbstractFormula;
import io.protobase.f7.formulas.FormulaName;
import io.protobase.f7.models.GridColumnRowKey;

import java.util.List;
import java.util.function.BiFunction;
import java.util.function.Function;

public class TYPE extends AbstractFormula {
  public static FormulaName NAME = FormulaName.TYPE;
  public static TYPE SELF = new TYPE();

  public TYPE() {
    super();
  }

  public TYPE(Function<Object, Object> lookup, BiFunction<GridColumnRowKey, Object, Object> collateralLookup) {
    super(lookup, collateralLookup);
  }

  @Override
  public Object internal(GridColumnRowKey origin, Object... args) {
    checkLength(args.length, 1, NAME);
    Object first = args[0];
    if (first instanceof Double) {
      return 1;
    }
    if (first instanceof String) {
      return 2;
    }
    if (first instanceof Boolean) {
      return 4;
    }
    if (first instanceof F7Exception) {
      return 16;
    }
    if (first instanceof List) {
      return 64;
    }
    return 128;
  }
}
